dask_expr/reductions.py

```diff
@@ -9,72 +11,141 @@
 )
 from dask.utils import M, apply
 
-from dask_expr.expr import Elemwise, Expr, Projection
+from dask_expr.expr import Blockwise, Elemwise, Expr, Projection
+
+##
+## Base MapReduce classes
+##
 
 
-class ApplyConcatApply(Expr):
-    """Perform reduction-like operation on dataframes
+class MapReduce(Expr):
+    """Perform a map-reduction operation on a dataframe
 
     This pattern is commonly used for reductions, groupby-aggregations, and
-    more.  It requires three methods to be implemented:
+    more. A `MapReduce` subclass must define the following attributes:
+
+    -   `map`: An expression class to use for the initial partition-wise
+        stage of the map-reduce. This should be a `Blockwise` expression
+        if possible.
+    -   `reduce`: An expression class to use for the reduction stage of
+        the map-reduce. This is typically a tree-reduction.
+
+    See Also
+    --------
+    Map
+    Reduce
+    DataFrameReduction
+    """
+
+    _parameters = ["frame"]
+    split_every: int = 0
+    map = None
+    reduce = None
+
+    @property
+    def _meta(self):
+        mapped = self.map(*self.operands)
+        reduced = self.reduce(mapped, *self.operands[1:])
+        return reduced._meta
+
+    def _divisions(self):
+        return (None, None)
+
+    def _simplify_down(self):
+        mapped = self.map(*self.operands)
+        return self.reduce(mapped, *self.operands[1:])
+
+
+class Map(Blockwise):
```

Comment:
Is this the same as `MapPartitions`?

I would expect `Map` to operate elementwise, not blockwise.  

I also expect us to need a `MapPartitions` operation (if we don't have one already).  Maybe that expression could do double duty?

Reply:
Unless we're looking to be explicit about differentiating reduction layers, then maybe `ReductionBlockwise`, `ReductionCombine`, ... or something similar